File: python/packages/ai/teams/ai/state/turn_state_manager.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from typing import Any, Generic, List, Optional, TypeVar

from botbuilder.core import Storage, TurnContext

from .turn_state import ConversationState, TempState, TurnState, UserState
from .turn_state_entry import TurnStateEntry

StateT = TypeVar("StateT", bound=TurnState)


class TurnStateManager(Generic[StateT]):
    "responsible for loading and saving an application turn state"

    async def load_state(self, storage: Optional[Storage], context: TurnContext) -> StateT:
        """
        loads all of the state scopes for the current turn\n
        `storage`: storage provider to load state scopes from\n
        `context`: context for the current turn of conversation with the user
        """

        if not context.activity:
            raise ValueError("missing context.activity")
        if not context.activity.channel_id:
            raise ValueError("missing context.activity.channel_id")
        if not context.activity.recipient:
            raise ValueError("missing context.activity.recipient")
        if not context.activity.recipient.id:
            raise ValueError("missing context.activity.recipient.id")
        if not context.activity.conversation:
            raise ValueError("missing context.activity.conversation")
        if not context.activity.conversation.id:
            raise ValueError("missing context.activity.conversation.id")
        if not context.activity.from_property:
            raise ValueError("missing context.activity.from_property")
        if not context.activity.from_property.id:
            raise ValueError("missing context.activity.from_property.id")

        channel_id = context.activity.channel_id
        conversation_id = context.activity.conversation.id
        bot_id = context.activity.recipient.id
        user_id = context.activity.from_property.id
        conversation_key = f"{channel_id}/{bot_id}/conversations/{conversation_id}"
        user_key = f"{channel_id}/{bot_id}/users/{user_id}"

        if storage:
            items: Any = await storage.read([conversation_key, user_key])
        else:
            items = {}

        conversation: ConversationState = (
            items[conversation_key] if conversation_key in items else ConversationState([])
        )
        user: UserState = items[user_key] if user_key in items else UserState()

        state: Any = TurnState(
            conversation=TurnStateEntry(conversation, storage_key=conversation_key),
            user=TurnStateEntry(user, storage_key=user_key),
            temp=TurnStateEntry(TempState(history="", input="", output="")),
        )

        return state

    async def save_state(self, storage: Optional[Storage], _context: TurnContext, state: StateT):
        "saves all of the state scopes for the current turn"

        if storage is None:
            return

        to_delete: List[str] = []
        changes = {}

        if state.conversation.storage_key:
            if state.conversation.is_deleted:
                to_delete.append(state.conversation.storage_key)
            elif state.conversation.has_changed:
                changes[state.conversation.storage_key] = state.conversation.value

        if state.user.storage_key:
            if state.user.is_deleted:
                to_delete.append(state.user.storage_key)
            elif state.user.has_changed:
                changes[state.user.storage_key] = state.user.value

        if state.temp.storage_key:
            if state.temp.is_deleted:
                to_delete.append(state.temp.storage_key)
            elif state.temp.has_changed:
                changes[state.temp.storage_key] = state.temp.value

        if len(changes) > 0:
            await storage.write(changes)

        if len(to_delete) > 0:
            await storage.delete(to_delete)

File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/State/TurnStateManager.cs
using Microsoft.Teams.AI.Exceptions;
using Microsoft.Teams.AI.Utilities;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Builder;

namespace Microsoft.Teams.AI.State
{
    /// <summary>
    /// Handles loading and saving of the configured turn state class.
    /// </summary>
    /// <typeparam name="TState">Optional. Type of turn state that encompasses the conversation, user and temp state.</typeparam>
    /// <typeparam name="TConversationState">Optional. Type of the conversation state object being persisted.</typeparam>
    /// <typeparam name="TUserState">Optional. Type of the user state object being persisted.</typeparam>
    /// <typeparam name="TTempState">Optional. Type of the temp state object being persisted.</typeparam>
    public class TurnStateManager<TState, TConversationState, TUserState, TTempState> : ITurnStateManager<TState>
        where TState : TurnState<TConversationState, TUserState, TTempState>, new()
        where TConversationState : StateBase, new()
        where TUserState : StateBase, new()
        where TTempState : TempState, new()
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="TurnStateManager{TState, TConversationState, TUserState, TTempState}"/> class.
        /// </summary>
        public TurnStateManager() { }

        /// <inheritdoc />
        public async Task<TState> LoadStateAsync(IStorage? storage, ITurnContext turnContext)
        {
            try
            {
                // Compute state keys
                Activity activity = turnContext.Activity;
                string channelId = activity.ChannelId;
                string botId = activity.Recipient.Id;
                string conversationId = activity.Conversation.Id;
                string userId = activity.From.Id;

                Verify.ParamNotNull(activity, "TurnContext.Activity");
                Verify.ParamNotNull(channelId, "TurnContext.Activity.ChannelId");
                Verify.ParamNotNull(botId, "TurnContext.Activity.Recipient.Id");
                Verify.ParamNotNull(conversationId, "TurnContext.Activity.Conversation.Id");
                Verify.ParamNotNull(userId, "TurnContext.Activity.From.Id");

                string conversationKey = $"{channelId}/${botId}/conversations/${conversationId}";
                string userKey = $"{channelId}/${botId}/users/${userId}";

                // read items from storage provider (if configured)
                IDictionary<string, object> items;
                if (storage != null)
                {
                    items = await storage.ReadAsync(new string[] { conversationKey, userKey });
                }
                else
                {
                    items = new Dictionary<string, object>();
                }

                TState state = new();
                TUserState? userState = null;
                TConversationState? conversationState = null;

                if (items.TryGetValue(userKey, out object userStateValue))
                {
                    userState = userStateValue as TUserState;
                }

                if (items.TryGetValue(conversationKey, out object conversationStateValue))
                {
                    conversationState = conversationStateValue as TConversationState;
                }

                userState ??= new TUserState();
                conversationState ??= new TConversationState();

                state.UserStateEntry = new TurnStateEntry<TUserState>(userState, userKey);
                state.ConversationStateEntry = new TurnStateEntry<TConversationState>(conversationState, conversationKey);
                state.TempStateEntry = new TurnStateEntry<TTempState>(new());

                return state;

            }
            catch (Exception ex)
            {
                throw new TeamsAIException($"Something went wrong when loading state: {ex.Message}", ex);
            }
        }

        /// <inheritdoc />
        public async Task SaveStateAsync(IStorage? storage, ITurnContext turnContext, TState turnState)
        {
            try
            {
                Verify.ParamNotNull(turnContext);
                Verify.ParamNotNull(turnState);

                Dictionary<string, object> changes = new();
                List<string> deletions = new();

                foreach (string key in turnState.Keys)
                {
                    if (turnState.TryGetValue(key, out IReadOnlyEntry<object> entry))
                    {
                        if (entry.StorageKey != null)
                        {
                            if (entry.IsDeleted)
                            {
                                deletions.Add(entry.StorageKey);
                            }

                            if (entry.HasChanged)
                            {
                                changes[entry.StorageKey] = entry.Value;
                            }
                        }
                    }

                }

                // Do we have a storage provider?
                if (storage != null)
                {
                    // Apply changes
                    List<Task> tasks = new();
                    if (changes.Keys.Count > 0)
                    {
                        tasks.Add(storage.WriteAsync(changes));
                    }

                    if (deletions.Count > 0)
                    {
                        tasks.Add(storage.DeleteAsync(deletions.ToArray()));
                    }

                    if (tasks.Count > 0)
                    {
                        await Task.WhenAll(tasks.ToArray());
                    }
                }
            }
            catch (Exception ex)
            {
                throw new TeamsAIException($"Something went wrong when saving state: {ex.Message}", ex);
            }
        }
    }

    /// <summary>
    /// Default turn state manager.
    /// </summary>
    public class TurnStateManager : TurnStateManager<TurnState, StateBase, StateBase, TempState>
    {
    }
}
